academy/4-my-own-chain/server-side.md

```diff
@@ -0,0 +1,1046 @@
+
+Go to the function and update it:
+
+```typescript [https://github.com/cosmos/academy-checkers-ui/blob/d623ae0/src/server/indexer.ts#L91-L93]
+const handleBlock = async (block: Block) => {
+    ...
+    const events: StringEvent[] = await client.getEndBlockEvents(block.header.height)
+    if (0 < events.length) console.log("")
+    await handleEvents(events)
+}
+```
+
+The events that you have converted are compatible with those emanating from transactions, so you can just pass them on. You still need to update `handleEvent` so that it acts on the new event type:
+
+```typescript [TODO]
+const handleEvent = async (event: StringEvent): Promise<void> => {
+    ...
+    if (isActionOf("GameForfeited")) {
+        // Function yet to be declared
+        await handleEventForfeit(event)
+    }
+}
+```
+
+To achieve this, add a new function:
+
+```typescript [https://github.com/cosmos/academy-checkers-ui/blob/d623ae0/src/server/indexer.ts#L196-L214]
+const handleEventForfeit = async (event: StringEvent): Promise<void> => {
+    const forfeitedId: string | undefined = getAttributeValueByKey(event.attributes, "IdValue")
+    if (!forfeitedId) throw new Error(`Forfeit event missing forfeitedId`)
+    const winner: string | undefined = getAttributeValueByKey(event.attributes, "Winner")
+    const blackAddress: string | undefined = db.games[forfeitedId]?.blackAddress
+    const redAddress: string | undefined = db.games[forfeitedId]?.redAddress
+    const blackGames: string[] = db.players[blackAddress]?.gameIds ?? []
+    const redGames: string[] = db.players[redAddress]?.gameIds ?? []
+    console.log(
+        `Forfeit game: ${forfeitedId}, black: ${blackAddress}, red: ${redAddress}, winner: ${winner}`,
+    )
+    const indexInBlack: number = blackGames.indexOf(forfeitedId)
+    if (0 <= indexInBlack) blackGames.splice(indexInBlack, 1)
+    const indexInRed: number = redGames.indexOf(forfeitedId)
+    if (0 <= indexInRed) redGames.splice(indexInRed, 1)
+    if (winner === "NO_PLAYER") {
+        delete db.games[forfeitedId]
+    }
+}
+```
+
+* Again there is a lot of error handling.
+* `handleEvent` deletes the game only if there are no winners, which means that it was a deletion, not a forfeit.
+
+## Test time of forfeit
+
+Run the previous tests again. Create a game and see how the deletion event is picked up:
+
+```
+Forfeit game: 1, black: cosmos1ac6srz8wh848zc08wrfghyghuf5cf3tvd45pnw, red: cosmos1t88fkwurlnusf6agvptsnm33t40kr4hlq6h08s, winner: NO_PLAYER
+```
+
+## Patch a game
+
+In the actions that the Express server exposes, `app.patch` must still be implemented. This allows a user to inform the server that its database is no longer synchronized, and that it should look at a specific game. It is a matter of data re-synchronization:
+
+1. If the game can be found in the blockchain state, update the indexer's database accordingly:
+    1. If there is a winner, then the game should be removed from its players' lists of games.
```

Comment:
Apologies if I am obviously wrong, but: are the statements in 1.1 and 1.2 correct here? Alternate:

1.1. If there is a winner, then the game should be ADDED TO its players' lists of games.
1.2. If there is no winner, then the game should be REMOVED FROM its players' lists of games.

That seems like the logical association to me, but I am only evaluating the language strings, and perhaps from the code's intention the existing text is correct.

Reply:
The existing text is correct. The objective of the code is to maintain a list of games on which the players have to play. So:

1. If there is no winner, it means it is a game that needs to be played on -> it needs to be presented to its players -> include it in the list of games.
2. If there is a winner, then the game is finished, nobody is expected to play on it -> remove it from the list of games.

Reply:
Understood, and thanks (also, nice to "meet" you!)